lib/Sema/CSDiag.cpp

```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
 }
 
 bool FailureDiagnosis::visitIfExpr(IfExpr *IE) {
-  auto typeCheckClauseExpr = [&](Expr *clause) -> Expr * {
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertTypePurpose) -> Expr * {
```

Comment:
@mkchoi212 @CodaFi is right, if we want to provide contextual type to `typeCheckClauseExpr` we need to use both type and its purpose from the original constraint system. It's just so happens that in the snippet attached to the patch it is CTP_ResultStmt, but it might easily be e.g. CTP_AssignSource. 

To properly fix the problem at hand (and you should also add additional test-case e.g. `let _: E = lhs < rhs ? .LessThan : .GreaterTham`) you'll have to make changes to `diagnoseMemberFailures` method which should be called when `.GreaterTham` is type-checked separately, since it doesn't matter what is the purpose of the contextual type when we use to to resolve `UnresolvedMemberExpr` it would work all different kinds of expressions.